rfcs/syntax-static-imports.md

```diff
@@ -0,0 +1,281 @@
+local New, Children, OnEvent, OnChange, Out, Ref, Cleanup = Fusion.New, Fusion.Children, Fusion.OnEvent, Fusion.OnChange, Fusion.Out, Fusion.Ref, Fusion.Cleanup
+local Tween, Spring = Fusion.Tween, Fusion.Spring
+```
+
+Previously, these issues were solved using `getfenv()`, `_G` or other similar features to try and 'inject' commonly used values from other modules. Since these are not good practice to depend upon - and can even lead to deoptimisation scenarios - there is a void left unfilled for codebases that previously would depend on these features for ergonomic reasons.
+
+The primary issue the above code block points to, is the lack of utilites available for destructuring imports. RFCs exist attempting to implement general value destructuring, but they are running into difficulties and ultimately could very well be too difficult to implement cleanly without sacrificing the conciseness that destructuring is pursued for in the first place. It should be noted that a likely-prominent use case for destructuring would be for extracting specific members from imported modules in a convenient way.
+
+The hope is, by exploring systems beyond `require()`, we can open up space to explore these ideas in a way that can preserve backwards compatibility much more easily, without ruining the potential for conciseness by having to work around existing syntax quirks.
+
+## Design
+
+This RFC proposes the addition of a static "import statement" to supersede almost all of the current use of `require()` in static contexts. *Almost* all, because there are many valid use cases for wanting to dynamically import modules, and these use cases are outside the scope of this change.
+
+```Lua
+!import "foo/bar/baz"
+```
+
+The `!import` used here is designed to mirror the current `export` used for exporting type annotations. To avoid ambiguity with function call syntax, the exclamation point is added, which lines up in style with the current `--!strict` style. This syntax may perhaps be extended in the future to cover other types of static declaration - perhaps even the aforementioned type checking mode.
+
+Unlike `require()` it is intended to introduce members to the namespace statically, not to return a value dynamically. This is why it is specifically a statement, not an expression:
+
+```Lua
+local foo = doSomething(require("foo")) -- ok
+local bar = doSomething(!import "foo") -- not ok
+```
+
+Since this statement is statically evaluated, the argument *must* be statically evaluatable (as is currently done to provide typechecking for `require()`):
+
+```Lua
+!import "foo" -- ok
+
+local bar = "foo"
+!import bar -- ok
+
+local bar = tostring(os.clock())
+!import bar -- not ok
+```
+
+The reason this is not specifically limited to string literals is to allow Roblox-like environments to evaluate statements for their imports, including the use of previously defined constant-like values:
+
+```Lua
+!import script.Parent.Libraries.Fusion
```

Comment:
Considering you want this to be completely static, what is the behavior of a script using this and being moved to a different location before running?

Reply:
This is a good question, to which I would say: I have probably used the wrong word here. When I say 'static', I mean 'the script won't dynamically change this', and I do not mean 'this should be baked into the bytecode'.

It should be evaluated at runtime, as `require()` is now.

Reply:
I have updated the RFC.